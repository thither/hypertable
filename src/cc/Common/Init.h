/*
 * Copyright (C) 2007-2016 Hypertable, Inc.
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 3
 * of the License, or any later version.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Hypertable. If not, see <http://www.gnu.org/licenses/>
 */

/** @file
 * Initialization helper for applications.
 *
 * The initialization routines help setting up application boilerplate,
 * configuration policies and usage strings.
 */

#ifndef Common_Init_h
#define Common_Init_h

#include <Common/Config.h>
#include <Common/System.h>

namespace Hypertable { namespace Config {

  /** @addtogroup Common
   *  @{
   */

  /**
   * Init with policy (with init_options (called before parse_args) and
   * init (called after parse_args) methods. The Policy template class is
   * usually a list of configuration policies, i.e.
   *
   *    typedef Meta::list<Policy1, Policy2, ...> MyPolicy;
   *
   * @sa Config::parse_args for params.
   *
   * @param argc The argc parameter of the main() function
   * @param argv The argv parameter of the main() function
   * @param desc Optional command option descriptor
   */
  template <class PolicyT>
  inline void init_with_policy(int argc, char *argv[], Desc *desc = 0) {
    try {
      System::initialize();

      std::lock_guard<std::recursive_mutex> lock(rec_mutex);
      properties = std::make_shared<Properties>();

      if (desc)
        cmdline_desc(*desc);

      PolicyT::init_options();
      parse_args(argc, argv);
      PolicyT::init();

      if (get<gBool>("verbose"))
        properties->print(std::cout);
    }
    catch (Exception &e) {
      PolicyT::on_init_error(e);
    }
  }

  /** Convenience function (more of a demo) to init with a list of polices
   * @sa init_with
   *
   * @param argc The argc parameter of the main() function
   * @param argv The argv parameter of the main() function
   * @param desc Optional command option descriptor
   */
  template <class PolicyListT>
  inline void init_with_policies(int argc, char *argv[], Desc *desc = 0) {
    typedef typename Join<PolicyListT>::type Combined;
    init_with_policy<Combined>(argc, argv, desc);
  }

  /** Initialize with default policy
   *
   * @param argc The argc parameter of the main() function
   * @param argv The argv parameter of the main() function
   * @param desc Optional command option descriptor
   */
  inline void init(int argc, char *argv[], Desc *desc = NULL) {
    init_with_policy<DefaultPolicy>(argc, argv, desc);
  }

  /** @} */

}}

#endif /* Common_Init_h */
