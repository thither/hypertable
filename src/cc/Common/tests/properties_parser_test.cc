/**
 * Copyright (C) 2007-2016 Hypertable, Inc.
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 3
 * of the License, or any later version.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

#include "Common/Compat.h"
#include "Common/Init.h"

using namespace Hypertable;

namespace {

struct AppPolicy : Config::Policy {
  static void init_options() {
	Config::cmdline_desc("Usage: %s [Options] [args]\nOptions").add_options()
   ("Hypertable.Config.OnFileChange.Reload2, OnFileChange.Reload2", cfg(false),
    "Set Config File Listener for Reloading cfg file on Change")
   ("i16", i16(1), "16-bit integer")
   ("i32", i32(), "32-bit integer")
   ("i64", i64(1), "64-bit integer")
   ("int64s", i64s(), "a list of 64-bit integers")
   ("boo,b", boo(false), "a boolean arg")
   ("string,str,s", str("A Default String Value"), "a string arg")
   ("strs", strs(), "a list of strings")
   ("float64", f64(3.12), "a double arg")
   ("f64s", f64s(), "a list of doubles")
   ("skip1", boo()->zero_token(), "a skipped token")
   ("skip2", "a skipped token")
   ("false", boo(false)->zero_token(), "token defaults to false")
   ("true", boo(true)->zero_token(), "token defaults to true")
   ("is_true", boo(false)->zero_token(), "if set token true")
   //("a.cmd.arg.qouted", str(), "a qouted string arg with spaces") require cmake escaping to testdiff  
   ;

	Config::cmdline_hidden_desc().add_options()
   ("args",   strs(), "rest of arguments")
   ("args",  -1 ) // -1, accumalate unknonw options
   ("app",    str(),  "application filename, zero pos skipped")
   ("app",    0)
   ("arg-1",  str(),  "argument one")
   ("arg-1",  1)
   ;
  }
};

typedef Meta::list<AppPolicy, Config::DefaultPolicy> Policies;

} // local namespace

int main(int argc, char *argv[]) {
  //Config::init_with_policies<Policies>(argc, argv);
  AppPolicy::init_options();
  
  std::cout << String("\nConfig::cmdline_desc()");
  std::cout << Config::cmdline_desc();
  std::cout << String("\nConfig::cmdline_hidden_desc()");
  std::cout << Config::cmdline_hidden_desc();

  Config::Parser prs(argc, argv, Config::cmdline_desc(), Config::cmdline_hidden_desc());
  
  std::cout << String("\n\nMerged New ParserConfig");
  std::cout << *prs.get_cfg();
  std::cout << prs; // Raw(Strings) Parsed
  prs.print_options(std::cout);
  

  // cfg file parse test definitions
  Config::file_desc().add_options()
   ("is_true_bool", boo(false), "a boolean arg")
   ("is_yes_bool", boo(false), "a boolean arg")
   ("is_one_bool", boo(false), "a boolean arg")
   ("is_no_bool", boo(true), "a boolean arg")
   ("is_rest_bool", boo(true), "a boolean arg")

   ("boo", boo(false), "a boolean arg")
   ("i16", i16(1), "16-bit integer")
   ("i32", i32(), "32-bit integer")
   ("i64", i64(1), "64-bit integer")
   ("int64s", i64s(), "a list of 64-bit integers")
   ("string", str("A Default String Value"), "a string arg")
   ("string.qouted", str("A Default String Value"), "a string arg")
   ("strs", strs(), "a list of strings")
   ("float64", f64(3.12), "a double arg")
   ("f64s", f64s(), "a list of doubles")
   ("missing", str("a missing string in cfg file, default applies"), "a string arg")
   ("aGroupOne.arg.1", i16(0), "a group arg")
   ("aGroupOne.arg.2", i16(0), "a group arg")
   ("aGroupOne.arg.3", i16(0), "a group arg")
   ("aGroupOne.arg.4", i16(0), "a group arg")
   ("aGroupOne.arg.5", i16(0), "a group arg")
    ;
    
  std::cout << String("\nConfig::file_desc()");
  std::cout << Config::file_desc();

  std::ifstream in("properties_parser_test.cfg");
  Config::Parser prs_file(in, Config::file_desc());
  
  std::cout << String("\nConfig::file_desc()");
  std::cout << Config::file_desc(); // original configuration (left untouched)

  std::cout << *prs_file.get_cfg();  // configurations options with values
  std::cout << prs_file; // Raw(Strings) Parsed
  prs_file.print_options(std::cout);
  
  Properties props;
  props.set("dummy", true);
  std::cout << props.str("dummy") << "\n";

  props.load_from(prs_file.get_options());

  props.print(std::cout, true);

}