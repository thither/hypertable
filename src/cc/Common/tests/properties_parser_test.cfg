
### a comment=argument

is_true_bool=True
is_yes_bool=yEs
is_one_bool=1
is_no_bool=false
    is_rest_bool=DefaultFalse  # followup comments
  # a_comment=withoutAmeaning

not_defined_name=True
[aGroupOne]
arg.1=1
  arg.2=2#comment
    arg.3=3  #comment
      arg.4=4  #comment
        arg.5=5  #comment
#comment
[aGroupOne=end]
boo         =            true
i16=64000
i32=640K
i64=10G
int64s=1G  
  int64s=2G  
    int64s=3G  
string=aStringNoSpaces
string.qouted   =  "a   S t r i n g   W i t h    S p a c e s"  #  comment

strs=StringOne  
  strs=StringTwo  
    strs=StringThree  

float64=1.1234
f64s=1.123
  f64s=1.123K
    f64s=1.123M
      f64s=1.123g


