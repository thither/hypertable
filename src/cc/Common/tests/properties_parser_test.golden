
Config::cmdline_desc()
Usage:  [Options] [args]
Options
  --Hypertable.Config.OnFileChange.Reload2 -[OnFileChange.Reload2]  Set Config File Listener for Reloading cfg file on Change  false
  --boo -[b]                                                        a boolean arg                                              false
  --f64s                                                            a list of doubles                                          []
  --false                                                           token defaults to false                                    false
  --float64                                                         a double arg                                               3.12
  --i16                                                             16-bit integer                                             1 
  --i32                                                             32-bit integer                                             0 
  --i64                                                             64-bit integer                                             1 
  --int64s                                                          a list of 64-bit integers                                  []
  --is_true                                                         if set token true                                          false
  --skip1                                                           a skipped token                                            true
  --skip2                                                           a skipped token                                            true
  --string -[str, s]                                                a string arg                                               A Default String Value
  --strs                                                            a list of strings                                          []
  --true                                                            token defaults to true                                     true

Config::cmdline_hidden_desc()

  --app    application filename, zero pos skipped    
  --arg-1  argument one                              
  --args   rest of arguments                       []


Merged New ParserConfig
Usage:  [Options] [args]
Options
  --Hypertable.Config.OnFileChange.Reload2 -[OnFileChange.Reload2]  Set Config File Listener for Reloading cfg file on Change  false
  --app                                                             application filename, zero pos skipped                       
  --arg-1                                                           argument one                                                 
  --args                                                            rest of arguments                                          []
  --boo -[b]                                                        a boolean arg                                              false
  --f64s                                                            a list of doubles                                          []
  --false                                                           token defaults to false                                    false
  --float64                                                         a double arg                                               3.12
  --i16                                                             16-bit integer                                             1 
  --i32                                                             32-bit integer                                             0 
  --i64                                                             64-bit integer                                             1 
  --int64s                                                          a list of 64-bit integers                                  []
  --is_true                                                         if set token true                                          false
  --skip1                                                           a skipped token                                            true
  --skip2                                                           a skipped token                                            true
  --string -[str, s]                                                a string arg                                               A Default String Value
  --strs                                                            a list of strings                                          []
  --true                                                            token defaults to true                                     true
*** Raw Parsed Options:
arg-1=[argOne]
boo=[true, 1]
i16=[1k]
i32=[64K]
i64=[1G]
is_true=[1]
strs=[1, 2, 3, 4, 5, 6]
*** Parsed Options:
Hypertable.Config.OnFileChange.Reload2=false
arg-1=argOne
boo=true
false=false
float64=3.12
i16=1000
i32=64000
i64=1000000000
is_true=true
string=A Default String Value
strs=[1, 2, 3, 4, 5, 6]
true=true

Config::file_desc()
Config Properties
  --aGroupOne.arg.1  a group arg                0 
  --aGroupOne.arg.2  a group arg                0 
  --aGroupOne.arg.3  a group arg                0 
  --aGroupOne.arg.4  a group arg                0 
  --aGroupOne.arg.5  a group arg                0 
  --boo              a boolean arg              false
  --f64s             a list of doubles          []
  --float64          a double arg               3.12
  --i16              16-bit integer             1 
  --i32              32-bit integer             0 
  --i64              64-bit integer             1 
  --int64s           a list of 64-bit integers  []
  --is_no_bool       a boolean arg              true
  --is_one_bool      a boolean arg              false
  --is_rest_bool     a boolean arg              true
  --is_true_bool     a boolean arg              false
  --is_yes_bool      a boolean arg              false
  --missing          a string arg               a missing string in cfg file, default applies
  --string           a string arg               A Default String Value
  --string.qouted    a string arg               A Default String Value
  --strs             a list of strings          []

Config::file_desc()
Config Properties
  --aGroupOne.arg.1  a group arg                0 
  --aGroupOne.arg.2  a group arg                0 
  --aGroupOne.arg.3  a group arg                0 
  --aGroupOne.arg.4  a group arg                0 
  --aGroupOne.arg.5  a group arg                0 
  --boo              a boolean arg              false
  --f64s             a list of doubles          []
  --float64          a double arg               3.12
  --i16              16-bit integer             1 
  --i32              32-bit integer             0 
  --i64              64-bit integer             1 
  --int64s           a list of 64-bit integers  []
  --is_no_bool       a boolean arg              true
  --is_one_bool      a boolean arg              false
  --is_rest_bool     a boolean arg              true
  --is_true_bool     a boolean arg              false
  --is_yes_bool      a boolean arg              false
  --missing          a string arg               a missing string in cfg file, default applies
  --string           a string arg               A Default String Value
  --string.qouted    a string arg               A Default String Value
  --strs             a list of strings          []

Config Properties
  --aGroupOne.arg.1  a group arg                0 
  --aGroupOne.arg.2  a group arg                0 
  --aGroupOne.arg.3  a group arg                0 
  --aGroupOne.arg.4  a group arg                0 
  --aGroupOne.arg.5  a group arg                0 
  --boo              a boolean arg              false
  --f64s             a list of doubles          []
  --float64          a double arg               3.12
  --i16              16-bit integer             1 
  --i32              32-bit integer             0 
  --i64              64-bit integer             1 
  --int64s           a list of 64-bit integers  []
  --is_no_bool       a boolean arg              true
  --is_one_bool      a boolean arg              false
  --is_rest_bool     a boolean arg              true
  --is_true_bool     a boolean arg              false
  --is_yes_bool      a boolean arg              false
  --missing          a string arg               a missing string in cfg file, default applies
  --string           a string arg               A Default String Value
  --string.qouted    a string arg               A Default String Value
  --strs             a list of strings          []
*** Raw Parsed Options:
aGroupOne.arg.1=[1]
aGroupOne.arg.2=[2]
aGroupOne.arg.3=[3]
aGroupOne.arg.4=[4]
aGroupOne.arg.5=[5]
boo=[true]
f64s=[1.123, 1.123K, 1.123M, 1.123g]
float64=[1.1234]
i16=[64000]
i32=[640K]
i64=[10G]
int64s=[1G, 2G, 3G]
is_no_bool=[false]
is_one_bool=[1]
is_rest_bool=[DefaultFalse]
is_true_bool=[True]
is_yes_bool=[yEs]
string=[aStringNoSpaces]
string.qouted=[a   S t r i n g   W i t h    S p a c e s]
strs=[StringOne, StringTwo, StringThree]
*** Parsed Options:
aGroupOne.arg.1=1
aGroupOne.arg.2=2
aGroupOne.arg.3=3
aGroupOne.arg.4=4
aGroupOne.arg.5=5
boo=true
f64s=[1.123, 1123, 1.123e+06, 1.123e+09]
float64=1.1234
i16=64000
i32=640000
i64=10000000000
int64s=[1000000000, 2000000000, 3000000000]
is_no_bool=false
is_one_bool=true
is_rest_bool=false
is_true_bool=true
is_yes_bool=true
missing=a missing string in cfg file, default applies
string=aStringNoSpaces
string.qouted=a   S t r i n g   W i t h    S p a c e s
strs=[StringOne, StringTwo, StringThree]
true
aGroupOne.arg.1=1
aGroupOne.arg.2=2
aGroupOne.arg.3=3
aGroupOne.arg.4=4
aGroupOne.arg.5=5
boo=true
dummy=true
f64s=[1.123, 1123, 1.123e+06, 1.123e+09]
float64=1.1234
i16=64000
i32=640000
i64=10000000000
int64s=[1000000000, 2000000000, 3000000000]
is_no_bool=false
is_one_bool=true
is_rest_bool=false
is_true_bool=true
is_yes_bool=true
missing=a missing string in cfg file, default applies (default)
string=aStringNoSpaces
string.qouted=a   S t r i n g   W i t h    S p a c e s
strs=[StringOne, StringTwo, StringThree]
